### ISSUE TITLE
S2583 FP: symbolic execution does not enter `do-while` loop

### ISSUE THREAD

**Issue by RalfKoban** [SonarSource/sonar-dotnet#2592]
### Description

SonarCube reports a S2583  warning when using pattern matching.

### Repro steps
Warning is reported for ```done is false```

Example code:
```C#
        public void Repro()
        {
            var done = false;
            do
            {
                done = true;
            }
            while (done is false);
        }
```

### Expected behavior
No warning given. This is a false positive.

### Actual behavior
Warning S2583 Change this condition so that it does not always evaluate to 'false'; some subsequent code is never executed.

### Known workarounds
Do not use pattern matching

### Related information

* SonarC# Version
7.16.0.8981

* Visual Studio Version
16.2.2

**Comment by andrei-epure-sonarsource** [comment]
thanks @RalfKoban  I confirm the FP. It's not specific to the `is` pattern but the `do-while` loop. The bellow is also a reproducer:

```
        public void Repro()
        {
            var done = false;
            do
            { // Secondary
                done = true;
            }
            while (done == false); // Noncompliant FP
        }
```

Our Symbolic Execution does not enter the loops and causes false positives.

We fixed `foreach` loops in #2524. We still have #890 open for `for` loops, and we now have your ticket for `do`-`while` loops.

Thanks again!

**Comment by mary-georgiou-sonarsource** [comment]
The issue that the user is reporting is a true positive.
`Do-while` and `is` work properly with the new SE engine.
See rule migration PR: #7750

**Comment by mary-georgiou-sonarsource** [comment]
The issue that the user is reporting is a true positive.
`Do-while` and `is` work properly with the new SE engine.
See rule migration PR: #7750
